internal/knowledge/query_sql.go

```diff
@@ -30,179 +31,274 @@ type SQLTranslation struct {
+			if scope.Context == WhereContext {
+				aliasPrefix = "rw"
+				assetAliasPrefix = "aw"
+			} else {
+				aliasPrefix = "r"
+				assetAliasPrefix = "a"
+			}
+			ralias = fmt.Sprintf("%s%d", aliasPrefix, relationCount)
+			relationCount++
+			var exps []string
 
-		if r.Direction == Right {
-			andExpressions.Children = append(andExpressions.Children, out)
-		} else if r.Direction == Left {
-			andExpressions.Children = append(andExpressions.Children, in)
-		} else if r.Direction == Either {
-			oneDirectionOptimization := false
-			// Optimization: in this case, finding in any direction is sufficient.
-			if len(queryGraph.Relations) == 1 {
-				nodesConstrained := false
-				for idx := range constrainedNodes {
-					if idx == r.LeftIdx {
-						nodesConstrained = true
-					}
-					if idx == r.RightIdx {
-						nodesConstrained = true
-					}
+			if len(relation.Labels) > 0 {
+				for _, label := range relation.Labels {
+					exps = append(exps, fmt.Sprintf("%s.type = '%s'", ralias, label))
 				}
-				if !nodesConstrained {
-					n, err := queryGraph.GetNodeByID(r.LeftIdx)
-					if err != nil {
-						return nil, nil, err
-					}
+			}
+
+			if relation.Direction == Right && relation.LeftIdx == i {
+				exps = append(exps, fmt.Sprintf("%s.from_id = %s%d.id", ralias, assetAliasPrefix, relation.LeftIdx))
+			} else if relation.Direction == Right && relation.RightIdx == i {
+				exps = append(exps, fmt.Sprintf("%s.to_id = %s%d.id", ralias, assetAliasPrefix, relation.RightIdx))
+			} else if relation.Direction == Left && relation.LeftIdx == i {
+				exps = append(exps, fmt.Sprintf("%s.to_id = %s%d.id", ralias, assetAliasPrefix, relation.LeftIdx))
+			} else if relation.Direction == Left && relation.RightIdx == i {
+				exps = append(exps, fmt.Sprintf("%s.from_id = %s%d.id", ralias, assetAliasPrefix, relation.RightIdx))
+			} else {
+				// If a relationship is undirected
+				optimize, err := isRelationOptimizable(queryGraph, *relation)
+
+				if err != nil {
+					return nil, nil, err
+				}
+
+				if !optimize { // if this relation is optimizable, then just on direction is needed as (v) -- (q) <==> (q) -- (v)
+					// if not, we need to fork a join to translate it into an UNION when building the SQL query
+					relationsCopy := make([]QueryRelation, len(queryGraph.Relations))
+					nodesCopy := make([]QueryNode, len(queryGraph.Nodes))
 
-					// if the node has a variable name bound to it, it is considered constrained
-					if len(n.Labels) > 0 {
-						nodesConstrained = true
+					copy(relationsCopy, queryGraph.Relations)
+					copy(nodesCopy, queryGraph.Nodes)
+
+					queryGraphClone := QueryGraph{
+						Nodes:          nodesCopy,
+						Relations:      relationsCopy,
+						VariablesIndex: queryGraph.VariablesIndex,
 					}
 
-					n, err = queryGraph.GetNodeByID(r.RightIdx)
+					// Calculate the join if the graph was directed to the right
+					queryGraphClone.Relations[relationsIdx[j]].Direction = Right
```

Comment:
idk how to feel about this `relationsIdx[j]` it will be very to break in the future

Reply:
Why? It's solely for getting the right element in the Relations array.

Do you think I should modify QueryRelation to have an index in its structure?

Reply:
I added the ID field to QueryNode and QueryRelation so it is easier to get a particular one directly from the graph, so no more `relationsIdx[j]`. I also updated QueryGraph tests to take this into account.

In case you're wondering, I didn't use the VariableIndex map field to do this because if the node doesn't have a variable name, then the index is not stored in the map in any way.